### PR TITLE
[move-ide] Added support for enums

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR adds enums support to the Move analyzer. 

In this PR, I also sneaked a fix for symbols merging, which was not very robust when multiple packages are built within the same IDE instance. It's much more robust to keep symbols on the per-package basis, even if it's more memory-intensive.

## Test plan 

All existing and newly added tests must pass.

### CODE REVIEW

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 1]
```diff
@@ -276,26 +314,26 @@ pub struct FieldDef {
...
-    pub field_defs: Vec<FieldDef>,
-    /// Does this struct have positional fields?
-    pub positional: bool,
-}
-
-impl StructDef {
-    pub fn name_start(&self) -> Position {
-        self.name_start
-    }
+pub enum MemberDefInfo {
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:318]: This change was done to be able to use one function (`add_member_use_def`) to add a `UseDef` for all relevant module members. Previously we had nearly identical functions to add them for functions and structs, and in this PR we'd need another one for enums

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 2]
```diff
@@ -1627,6 +1741,68 @@ fn ignored_function(name: Symbol) -> bool {
 
 /// Main AST traversal functions
 
+fn field_defs_and_types(
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1744]: Extracted to be shared between struct fields and variant field

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 3]
```diff
@@ -361,23 +356,61 @@ impl TypingAnalysisContext<'_> {
...
                 _,
             ) => {
-                self.add_field_use_def(mod_ident, &struct_name.value(), use_name, use_pos);
+                self.add_struct_field_use_def(mod_ident, &struct_name.value(), use_name, use_pos);
             }
             _ => (),
         }
     }
 
+    fn add_field_use_def_internal(
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:379]: Extracted to be shared by struct and variant field addition functions

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 4]
```diff
@@ -461,6 +562,77 @@ impl TypingAnalysisContext<'_> {
...
+
+        self.visit_type(None, &mut match_pat.ty);
+        match &mut match_pat.pat.value {
+            UA::Variant(mident, name, vname, tyargs, fields)
+            | UA::BorrowVariant(_, mident, name, vname, tyargs, fields) => {
+                self.add_datatype_use_def(mident, name);
+                self.add_variant_use_def(mident, name, vname);
+                tyargs.iter_mut().for_each(|t| self.visit_type(None, t));
+                for (fpos, fname, (_, (_, pat))) in fields.iter_mut() {
+                    if self.compiler_info.ellipsis_binders.get(&fpos).is_none() {
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:577]: @cgswords, I only use the location of identifiers that need to be "hidden" as they are part of (internally expanded ellipsis). From what I understand these identifiers get the same location as the ellipsis so it should work (and indeed does in practice, at least on examples I tried) but I wonder if you can see something that could break it

Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:577]: We don't really have location for ellipsis but I guess I could see if ANY identifier amongst binders is annotated with `EllipsisMatchEntries`. One problem is that I don't think we can tell at naming if a given variant has positional or named fields so it's unclear which variant of the `EllipsisMatchEntries` enum to check against.. Hence the idea of checking location of each identifier for the IDE annotation

Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:577]: I guess my thinking is that we would want to associate the pattern with this entry so that we could offer users to "expand" the ellipses. That won't help us much in the positional case, however, and we'll still need to do something like this.

Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:577]: > I guess my thinking is that we would want to associate the pattern with this entry so that we could offer users to "expand" the ellipses.

This may not be a bad idea to show ellipsis expansion when hovering over the wild card (not in this PR, though, I think). I will keep the current "hidden" identifiers handling as is then unless I misunderstood and you do like me to change it

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 5]
```diff
@@ -795,6 +877,44 @@ fn fun_type_to_ide_string(fun_type: &FunType) -> String {
     .to_string()
 }
 
+fn abilities_to_id_string(abilities: &AbilitySet) -> String {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:880]:
```suggestion
fn abilities_to_ide_string(abilities: &AbilitySet) -> String {
```

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 6]
```diff
@@ -2851,103 +3152,124 @@ pub fn on_use_request(
...
-        let detail = Some(mod_def.ident.clone().to_string());
-        let kind = SymbolKind::MODULE;
-        let Some(range) = symbols.files.lsp_range_opt(&mod_def.name_loc) else {
-            continue;
-        };
-
-        let mut children = vec![];
-
-        // handle constants
-        let cloned_const_def = mod_def.constants.clone();
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:2876]: Cloning here and other similar places was unnecessary so I fixed it while I was modifying these functions to support enums.

File: external-crates/move/crates/move-analyzer/src/analyzer.rs [hunk 7]
```diff
@@ -44,7 +44,7 @@ pub fn run() {
     );
 
     let (connection, io_threads) = Connection::stdio();
-    let symbols = Arc::new(Mutex::new(symbols::empty_symbols()));
+    let symbols_map = Arc::new(Mutex::new(BTreeMap::new()));
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/analyzer.rs:47]: This change is not directly related to enums support but enums support would not work if we did not merge `CompilerInfo` at the end of symbols computation. When thinking about it, I realized that merging is not the greatest idea after all and figured I should fix it

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 8]
```diff
@@ -65,13 +62,18 @@ pub struct TypingAnalysisContext<'a> {
...
     /// definition already.
-    fn file_start_position_opt(&self, loc: &Loc) -> Option<files::FilePosition> {
-        self.files.file_start_position_opt(loc)
+    fn file_start_position_opt(&self, loc: &Loc) -> Option<Position> {
+        self.files
+            .file_start_position_opt(loc)
+            .map(|p| lsp_types::Position {
+                line: p.position.line_offset() as u32,
+                character: p.position.column_offset() as u32,
+            })
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:71]:
```suggestion
        self.files
            .file_start_position_opt(loc)
            .map(|p| p.into())
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 9]
```diff
@@ -533,10 +534,19 @@ pub fn on_completion_request(
...
+                        completion_items(pos, &path, symbols)
+                    } else {
+                        vec![]
+                    }
+                }
             }
         }
-        None => completion_items(pos, &path, &context.symbols.lock().unwrap()),
+        None => {
+            eprintln!("failed completion for {:?} (package root not foune)", path);
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:547]:
```suggestion
            eprintln!("failed completion for {:?} (package root not found)", path);
```

File: external-crates/move/crates/move-analyzer/src/inlay_hints.rs [hunk 10]
```diff
@@ -25,43 +25,45 @@ pub fn on_inlay_hint_request(context: &Context, request: &Request, symbols: &Sym
...
-                            padding_left: None,
-                            padding_right: None,
-                            data: None,
-                        };
-                        hints.push(h);
+    if let Some(symbols) =
+        SymbolicatorRunner::root_dir(&fpath).and_then(|pkg_path| symbols_map.get(&pkg_path))
+    {
+        if context.inlay_type_hints {
+            if let Some(file_defs) = symbols.file_mods.get(&fpath) {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/inlay_hints.rs:32]: It would be a lot cleaner to pull this out into a function that returns `Option<Vec<InlayHint>>` so you could do these as serial `let` statements with `?` at the end of each.

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 11]
```diff
@@ -2173,9 +2404,36 @@ impl<'a> ParsingSymbolicator<'a> {
...
+                v.iter().for_each(|e| {
+                    if let P::Ellipsis::Binder((_, m)) = e {
+                        self.match_pattern_symbols(m);
+                    }
+                })
+            }
+            MP::Name(_, chain) => self.chain_symbols(chain),
+            MP::Or(m1, m2) => {
+                self.match_pattern_symbols(m1);
+                self.match_pattern_symbols(m2);
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:2433]: Small nit: we may prefer to process `m2` first so the def location is the soonest one possible.

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 12]
```diff
@@ -2173,9 +2404,36 @@ impl<'a> ParsingSymbolicator<'a> {
...
+            }
+            MP::At(_, m) => self.match_pattern_symbols(m),
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:2435]: Shouldn't we process the binding for the `At`?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 13]
```diff
@@ -2750,20 +3011,50 @@ pub fn on_references_request(context: &Context, request: &Request, symbols: &Sym
...
+    if !arm_binder {
+        return None;
+    }
+    let Some(fhash) = symbols.file_hash(use_fpath) else {
+        return None;
+    };
+    let Some(byte_idx) = lsp_position_to_byte_index(&symbols.files, fhash, position) else {
+        return None;
+    };
+    let loc = Loc::new(fhash, byte_idx, byte_idx);
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:3044]:
```suggestion
    let DefInfo::Local(name, ty, is_let, is_mut, arm_binder) = def_info else {
        return None;
    };
    if !arm_binder {
        return None;
    }
    let fhash = symbols.file_hash(use_fpath)?;
    let loc = lsp_position_to_loc(&symbols.files, fhash, position)?;
```

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 14]
```diff
@@ -2750,20 +3011,50 @@ pub fn on_references_request(context: &Context, request: &Request, symbols: &Sym
...
+    use_fpath: &Path,
+    position: &Position,
+    symbols: &Symbols,
+) -> Option<DefInfo> {
+    let DefInfo::Local(name, ty, is_let, is_mut, arm_binder) = def_info else {
+        return None;
+    };
+    if !arm_binder {
+        return None;
+    }
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:3037]: Could we just store the guard's `Option<Loc>` in this field?

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 15]
```diff
@@ -437,6 +542,105 @@ impl TypingAnalysisContext<'_> {
...
+                self.process_match_patterm(pat2);
+            }
+            UA::At(var, pat) => {
+                self.add_local_def(
+                    &var.loc,
+                    &var.value.name,
+                    match_pat.ty.clone(),
+                    false,
+                    false,
+                    false,
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:600]: Why is this false?

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 16]
```diff
@@ -437,6 +542,105 @@ impl TypingAnalysisContext<'_> {
...
+                }
+            }
+            UA::Constant(mod_ident, name) => self.add_const_use_def(mod_ident, name),
+            UA::Binder(mut_, var) => self.add_local_def(
+                &var.loc,
+                &var.value.name,
+                match_pat.ty.clone(),
+                false,
+                matches!(mut_, E::Mutability::Mut(_)),
+                false,
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:587]: Why is this false?

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 17]
```diff
@@ -437,6 +542,105 @@ impl TypingAnalysisContext<'_> {
...
+            }
+            UA::Literal(_) | UA::Wildcard | UA::ErrorPat => (),
+        }
+    }
+
+    fn process_match_arm(&mut self, sp!(_, arm): &mut T::MatchArm) {
+        self.process_match_patterm(&mut arm.pattern);
+        arm.binders.iter_mut().for_each(|(var, ty)| {
+            self.add_local_def(&var.loc, &var.value.name, ty.clone(), false, false, true);
+        });
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:612]: Do we need to add the locals for the patterns during pattern processing if we're also doing it here?

**Review by cgswords — CHANGES_REQUESTED** [review]
Mostly just small nits left. It's start to look very nice!

File: external-crates/move/crates/move-analyzer/src/inlay_hints.rs (hunk 10 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/inlay_hints.rs:32]: Good idea!

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 14 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs]: Yes. I was not sure it was necessary but after our recent conv I now understand why it's needed

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 12 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:2436]: There is no action in the parsing symbolicator for variable definitions

File: external-crates/move/crates/move-analyzer/src/inlay_hints.rs [hunk 18]
```diff
@@ -74,6 +35,45 @@ pub fn on_inlay_hint_request(context: &Context, request: &Request, symbols: &Sym
     }
 }
 
+fn inlay_hints(context: &Context, fpath: PathBuf) -> Option<Vec<InlayHint>> {
+    let symbols_map = &context.symbols.lock().unwrap();
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/inlay_hints.rs:39]:
```suggestion
    let symbols_map = &context.symbols.lock()?;
```
Or if `lock()` is a result, 
```suggestion
    let symbols_map = &context.symbols.lock().ok()?;
```

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 19]
```diff
@@ -713,32 +981,53 @@ impl<'a> TypingVisitorContext for TypingAnalysisContext<'a> {
...
+                        .for_each(|t| visitor.visit_type(Some(exp_loc), t));
                     true
                 }
                 TE::Borrow(_, exp, field) => {
                     visitor.visit_exp(exp);
-                    visitor.add_field_type_use_def(&exp.ty, &field.value(), &field.loc());
+                    visitor.add_struct_field_type_use_def(&exp.ty, &field.value(), &field.loc());
+                    true
+                }
+                TE::PackVariant(mident, name, vname, tyargs, fields) => {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:1004]: This _should_ be unreachable.

**Review by cgswords — APPROVED** [review]
Awesome! LGTM!

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs (hunk 19 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:1004]: If we `debug_assert` it to `false` (much like we do for `TE::VariantMatch` right below), the current tests will fail on this assertion
